core/src/main/java/com/sequenceiq/cloudbreak/service/recipe/UpdateRecipeService.java

```diff
@@ -86,8 +91,22 @@ private UpdateHostGroupRecipesPair doHostGroupRecipeUpdate(Map<String, Set<Strin
                         .collect(Collectors.toSet());
                 attachHostGroupRecipesOpt = collectAttachHostGroupRecipes(hostGroupName, existingRecipeNames, updateRecipes);
                 detachHostGroupRecipesOpt = collectDetachHostGroupRecipes(hostGroupName, recipesForHostGroup, existingRecipes);
+                Set<GeneratedRecipe> generatedRecipeToDeleteSet = new HashSet<>();
+                if (detachHostGroupRecipesOpt.isPresent()) {
+                    UpdateHostGroupRecipes deletedHostGroupRecipes = detachHostGroupRecipesOpt.get();
+                    Set<GeneratedRecipe> newGeneratedRecipes = hostGroup.getGeneratedRecipes().stream()
+                            .filter(gr -> !recipeNamesMatchForGeneratedRecipe(gr, deletedHostGroupRecipes.getRecipeNames()))
+                            .collect(Collectors.toSet());
+                    generatedRecipeToDeleteSet = hostGroup.getGeneratedRecipes().stream()
+                            .filter(gr -> recipeNamesMatchForGeneratedRecipe(gr, deletedHostGroupRecipes.getRecipeNames()))
+                            .collect(Collectors.toSet());
+                    hostGroup.setGeneratedRecipes(newGeneratedRecipes);
```

Comment:
I think this could be simplified: 
`hostGroup.getGeneratedRecipes().removeAll(generatedRecipeToDeleteSet);`